### DISCUSSION TITLE
Do not clone repositories when there is no config and onboarding is disabled

### DISCUSSION THREAD

**Discussion by aentwist** [renovatebot/renovate#25672] [Suggest an Idea]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

37.43.4

### If you're self-hosting Renovate, select which platform you are using.

GitLab self-hosted

### Was this something which used to work for you, and then stopped?

No

### Wanted end result.

I am running renovate with autodiscover and without onboarding. I want it to only run on repos that are configured / enabled. I expect renovate to determine whether repos are enabled before starting work on them. I do not want it to clone repos that are not enabled as this is very expensive, but I am under the impression that is not currently the case. How can I set cloning to be lazy?

### What you tried so far.

Having trouble finding info on this. Haven't tried anything, open to suggestions.

### Relevant debug logs

For example

```
INFO: Repository is disabled - skipping (repository=andersonentwistle/sticky-scroll)
 INFO: Repository finished (repository=andersonentwistle/sticky-scroll)
       "cloned": true,
       "durationMs": 1066
```

There are many repos, and some take a full 2 minutes to be cloned.

**Comment by rarkins** [comment]
Renovate supports a lot of config file names, so you'd need to limit yourself to one (eg renovate.json) to avoid doing too many API calls.  Then there could be a variant if optimizeForDisabled adapted for the case where onboarding=false

**Reply by aentwist** [reply]
Oh wow thanks for pointing out that config option. That is definitely the closest thing available. But as you also point out, I guess this is a feature request.

From `optimizeForDisabled` we have,

> If the file exists and the config is disabled, Renovate will skip the repo without cloning it.

which should become,

> If **the file does not exist and onboarding is disabled or** the file exists and the config is disabled, Renovate will skip the repo without cloning it.

This should not be a breaking change, fortunately? What do you think?

---

Here is another feature request, because you also point out that the config file must have a particular name for this to work. This is also documented at `optimizeForDisabled`.

> Try to fetch the default config file (e.g. renovate.json)
[...]
If onboardingConfigFileName is set, that file name will be used instead of the default.

Some platforms appear to allow searching a "tree". This would mean you could get data on whether any Renovate config files exist without making a request for each one. Ideally I don't want to require a specific config file name to get optimization benefits. So make the (platform-specific) tree request and match it with a config file name regex if the platform supports it, otherwise fall back to the existing behavior.

###### GitHub

> To get a repository's contents recursively, you can [recursively get the tree](https://docs.github.com/rest/git/trees#get-a-tree).

https://docs.github.com/en/rest/repos/contents?apiVersion=2022-11-28#get-repository-content
-> https://docs.github.com/en/rest/git/trees?apiVersion=2022-11-28#get-a-tree

###### GitLab

https://docs.gitlab.com/ee/api/repositories.html#list-repository-tree

**Reply by aentwist** [reply]
I have no problem with requiring a specific config file name if we focus on # 1 at first. It is a much lower hanging fruit than the second idea with much higher impact.

this

https://github.com/renovatebot/renovate/blob/f4f7a899aed37d92440763dc8ced24808b997eb3/lib/workers/repository/init/apis.ts#L29-L36

needs this

```typescript
const renovateConfig = await getJsonFile(defaultConfigFile(config));
// +++
if (!renovateConfig && config.onboarding === false || renovateConfig?.enabled === false) {
// +++
  throw new Error(REPOSITORY_DISABLED_BY_CONFIG);
}
```

and we need another test, probably

https://github.com/renovatebot/renovate/blob/f4f7a899aed37d92440763dc8ced24808b997eb3/lib/workers/repository/init/apis.spec.ts#L32

and the docs

https://github.com/renovatebot/renovate/blob/f4f7a899aed37d92440763dc8ced24808b997eb3/lib/config/options/index.ts#L509-L511

https://github.com/renovatebot/renovate/blob/f4f7a899aed37d92440763dc8ced24808b997eb3/docs/usage/self-hosted-configuration.md?plain=1#L617

**Reply by rarkins** [reply]
That would be a change of behavior / breaking change and not ok.

Any new behavior would need to be based on additional logic triggered by onboarding=false

**Reply by aentwist** [reply]
I see... if onboarding is disabled and the config file doesn't exist but there is a config file by a different name, the repo would be skipped where it wasn't before. This wouldn't be the case if we read the tree with 1-2 requests and checked for all config files... so I guess that *is* actually the ask. That is not breaking... right? And if the breaking optimization were to be added after, then it would also not be breaking?

But that is a lot of work. So for now I guess the existing way to do this is to disable all in the global config and then reenable? If I understand that correctly then I will try that, because this will speed up my runs by an order of magnitude.

Just adding another config option is starting to sound pretty good 😅

---

By the way, this is a pretty weird piece of code

https://github.com/renovatebot/renovate/blob/f4f7a899aed37d92440763dc8ced24808b997eb3/lib/workers/repository/init/apis.ts#L47-L51

Note that if `renovateConfig?.extends?.includes(':enableRenovate') ??` evaluates to false (renovate config exists and has extends but extends doesn't include enableRenovate) then the nullish operator will evaluate to the LHS (false is not null or undefined) without evaluating the other conditions. Same story for the next condition. Not sure if that is intentional but I'd be amazed if so. Hard to read / maintain - I literally cannot tell what is happening there. Relying hard on that comment and trusting the documentation.

---

If I get some spare time I can try to look into requesting git trees and matching for config files if you are interested. No guarantees though.

**Reply by aentwist** [reply]
@rarkins I played around with the GitLab GraphQL API for an hour and came up with this:

```graphql
{
  project(fullPath: "andersonentwistle/vue-template") {
    repository {
      rootDir: tree {
        blobs {
          nodes {
            path
          }
        }
      }
      githubDir: tree(path: ".github") {
        blobs {
          nodes {
            path
          }
        }
      }
      gitlabDir: tree(path: ".gitlab") {
        blobs {
          nodes {
            path
          }
        }
      }
    }
  }
}
```

Which gives

```json
{
  "data": {
    "project": {
      "repository": {
        "rootDir": {
          "blobs": {
            "nodes": [
              {
                "path": ".commitlintrc.json"
              },
              {
                "path": ".eslintrc.cjs"
              },
              {
                "path": ".gitattributes"
              },
              {
                "path": ".gitignore"
              },
              {
                "path": ".gitlab-ci.yml"
              },
              {
                "path": ".pre-commit-config.yaml"
              },
              {
                "path": ".prettierrc.json"
              },
              {
                "path": ".releaserc.json"
              },
              {
                "path": ".renovaterc.json"
              },
              {
                "path": ".tool-versions"
              },
              {
                "path": "CONTRIBUTING.md"
              },
              {
                "path": "README.md"
              },
              {
                "path": "env.d.ts"
              },
              {
                "path": "index.html"
              },
              {
                "path": "package-lock.json"
              },
              {
                "path": "package.json"
              },
              {
                "path": "template.md"
              },
              {
                "path": "tsconfig.app.json"
              },
              {
                "path": "tsconfig.config.json"
              },
              {
                "path": "tsconfig.json"
              },
              {
                "path": "tsconfig.vitest.json"
              },
              {
                "path": "vite.config.ts"
              }
            ]
          }
        },
        "githubDir": {
          "blobs": {
            "nodes": []
          }
        },
        "gitlabDir": {
          "blobs": {
            "nodes": []
          }
        }
      }
    }
  }
}
```

Here are some notes:

- I have intentionally chosen GraphQL over REST here for the powerful request multiple resources and filtering capabilities
- I understand this could be a large request and limit the scope/bandwidth to minimize it
  - I limit the tree requests to root, .github, and .gitlab because that is where the config files live
  - I limit the data to the relative path `path`. We cannot just use `name` because only specific config files can live in the .github and .gitlab folders
  - I have not found a way to tell them to do the search on their side
- Optimizing for package.json config sections would require special handling to check whether those sections actually exist within package.json files

So vendor specific, do this, loop and match `path` to ... hm, this https://regex101.com/r/Cujd5Z/1. What do you think?

**Reply by aentwist** [reply]
Here is a crappy WIP query for GitHub

```graphql
{
  repository(owner: "aentwist", name: "pre-commit-mirrors-commitlint") {
    object(expression: "HEAD:") {
      ... on Tree {
        entries {
          path
        }
      }
    }
  }
}
```

https://docs.github.com/en/graphql/overview/explorer

**Reply by rarkins** [reply]
No GraphQL for this - it is much more heavily rate limited than REST

**Reply by aentwist** [reply]
Then it's just the same, except with 3 REST requests + over double the bandwidth on each. That is fine. I see there are nontrivial calculations to determine how much rate allowance each GitHub GQL request takes :( maybe GitLab is better for this, I'd hope.

**Comment by jamietanna** [comment]
[`optimizeForDisabled`](https://docs.renovatebot.com/self-hosted-configuration/#optimizefordisabled) was added for this

**Reply by aentwist** [reply]
As discussed in great detail above, optimizeForDisabled did not cover it. Maybe it has since been updated.
